### PR TITLE
Sets `CNB_PLATFORM_API` env var on lifecycle containers

### LINKED ISSUE #794: Use Platform API 0.4 when available
### Description

We should update `pack` to use the latest (and greatest) platform API (`0.4`)

### Proposed solution

Detect the supported platform APIs from the lifecycle and use 0.4 when available. Fallback to platform `0.3` when appropriate.

### Additional context
- [ ] Determines best Platform API version to use when lifecycle provides multiple
- [ ] Sets `CNB_PLATFORM_API` to `0.4` when executing the lifecycle container

### PR DESCRIPTION
**Pull request by jromero**
## Summary

Sets `CNB_PLATFORM_API` env var on lifecycle containers.

Also, refactored "Lifecycle" to separate the responsibility of "executor" and "execution".

## Documentation

- Should this change be documented?
    - Yes, [already documented as part of spec](https://github.com/buildpacks/spec/blob/platform/v0.4/platform.md#platform-api-compatibility)

## Related

Resolves #794

### CODE REVIEW

File: internal/build/lifecycle_execution.go [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+}
+
+func NewLifecycleExecution(logger logging.Logger, docker client.CommonAPIClient, opts LifecycleOptions) (*LifecycleExecution, error) {
+	latestSupportedPlatformAPI, err := findLatestSupported(append(
+		opts.Builder.LifecycleDescriptor().APIs.Platform.Deprecated,
+		opts.Builder.LifecycleDescriptor().APIs.Platform.Supported...,
+	))
+	if err != nil {
+		return nil, err
+	}
```
Comment (jromero) [inline internal/build/lifecycle_execution.go:39]: This is the main change, where we determine the right platform API version to use.

File: internal/build/phase_config_provider.go [hunk 2]
```diff
@@ -26,28 +27,29 @@ type PhaseConfigProvider struct {
...
 	provider.ctrConf.Labels = map[string]string{"author": "pack"}
 
-	if lifecycle.os == "windows" {
+	if lifecycleExec.os == "windows" {
 		provider.hostConf.Isolation = container.IsolationProcess
 	}
 
 	ops = append(ops,
-		WithLifecycleProxy(lifecycle),
+		WithEnv(fmt.Sprintf("%s=%s", platformAPIEnvVar, lifecycleExec.platformAPI.String())),
```
Comment (jromero) [inline internal/build/phase_config_provider.go:48]: This is where we set the new Env Var.

File: acceptance/acceptance_test.go [hunk 3]
```diff
@@ -476,6 +476,7 @@ func testWithoutSpecificBuilderRequirement(
 			output, err := pack.Run(
 				"build", "some/image",
 				"-p", filepath.Join("testdata", "mock_app"),
+				"--default-process", "web",
```
Comment (jromero) [inline acceptance/acceptance_test.go:479]: Adding of `"--default-process", "web",` is a temporary work around for an know lifecycle issue: https://github.com/buildpacks/lifecycle/issues/376

File: build.go [hunk 4]
```diff
@@ -40,7 +40,7 @@ const (
 	minLifecycleVersionSupportingImage   = "0.7.5"
 )
 
-type Lifecycle interface {
+type LifecycleExecutor interface {
```
Comment (dfreilich) [inline build.go:43]: 👍  to this rename

File: internal/build/lifecycle_executor.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	DefaultProcessType string
+	FileFilter         func(string) bool
+}
+
+func NewLifecycleExecutor(logger logging.Logger, docker client.CommonAPIClient) *LifecycleExecutor {
+	return &LifecycleExecutor{logger: logger, docker: docker}
+}
+
+func (l *LifecycleExecutor) Execute(ctx context.Context, opts LifecycleOptions) error {
+	lifecycleExec, err := NewLifecycleExecution(l.logger, l.docker, opts)
```
Comment (dfreilich) [inline internal/build/lifecycle_executor.go:72]: This is so much cleaner. Big fan.

**Pull request merged**